### PR TITLE
Ensure OGR geometries are cleaned up in Iterator._next()

### LINKED ISSUE #1204: Memory leak using items() iterator using step
Hi, 

I think I encountered a memory leak while using fiona

## Expected behavior and actual behavior.

I expected to be able to iterate trough a fiona Collection  created from a GeoPackage using the iterator defined calling the `items` method with a step defined.
Instead when I use the iterator, my memory becomes saturated. The bigger the step, the faster the memory usage increases.
Iterarating directly trough trough the Collection or with an iterator with no step does not seem to increase memory usage.

## Steps to reproduce the problem.

```
with fio.open("file.gpkg") as c:
    it = c.items(0, len(c), 2)
    for _, f it:
        p = f['properties']
```

## Operating system

Ubuntu 20.04.5 LTS

## Fiona and GDAL version and provenance

fiona 1.9.1
GDAL 3.0.4 installed via apt

### PR DESCRIPTION
**Pull request by sgillies**
Resolves #1204

### CODE REVIEW

**Comment by sgillies** [comment]
GDAL 3.5.3 fails to compile on the current Ubuntu 18.04 image https://github.com/Toblerity/Fiona/actions/runs/4212452737/jobs/7311484298#step:1:9. Image is described at https://github.com/actions/runner-images/blob/ubuntu18/20230217.2/images/linux/Ubuntu1804-Readme.md.

Last successful 3.5.3 build was this one: https://github.com/Toblerity/Fiona/actions/runs/4146922071/jobs/7173183429#step:1:9. An earlier image version (20230118.2).

File: scripts/travis_gdal_install.sh [hunk 1]
```diff
@@ -51,153 +51,156 @@ GDAL_CMAKE_OPTS=" -DBUILD_APPS=OFF \
...
-        -DOGR_ENABLE_DRIVER_SVG=OFF \ 
-        -DOGR_ENABLE_DRIVER_SXF=OFF \ 
-        -DOGR_ENABLE_DRIVER_TAB=ON \ 
-        -DOGR_ENABLE_DRIVER_TIGER=OFF \ 
-        -DOGR_ENABLE_DRIVER_VDV=OFF \ 
-        -DOGR_ENABLE_DRIVER_VFK=OFF \ 
-        -DOGR_ENABLE_DRIVER_VRT=ON \ 
-        -DOGR_ENABLE_DRIVER_WASP=OFF \ 
-        -DOGR_ENABLE_DRIVER_WFS=OFF \ 
-        -DOGR_ENABLE_DRIVER_XLS=OFF \
```
Comment (sgillies) [inline scripts/travis_gdal_install.sh:200]: I removed trailing whitespace from many of these lines.

File: scripts/travis_gdal_install.sh [hunk 2]
```diff
@@ -51,153 +51,156 @@ GDAL_CMAKE_OPTS=" -DBUILD_APPS=OFF \
...
-        -DOGR_ENABLE_DRIVER_TIGER=OFF \ 
-        -DOGR_ENABLE_DRIVER_VDV=OFF \ 
-        -DOGR_ENABLE_DRIVER_VFK=OFF \ 
-        -DOGR_ENABLE_DRIVER_VRT=ON \ 
-        -DOGR_ENABLE_DRIVER_WASP=OFF \ 
-        -DOGR_ENABLE_DRIVER_WFS=OFF \ 
-        -DOGR_ENABLE_DRIVER_XLS=OFF \ 
+        -DGDAL_USE_MSSQL_NCLI=OFF \
+        -DGDAL_USE_MSSQL_ODBC=OFF \
+        -DGDAL_USE_ODBC=OFF \
```
Comment (sgillies) [inline scripts/travis_gdal_install.sh:56]: Apparently the new GHA runner image added incomplete ODBC support and Cmake found it, but couldn't build it successfully. I feel like a lot of optional features should have to be explicitly asked for. The build breakage I'm fixing could happen again in other corners of GDAL.

**Comment by rouault** [comment]
> GDAL 3.5.3 fails to compile on the current Ubuntu 18.04 image https://github.com/Toblerity/Fiona/actions/runs/4212452737/jobs/7311484298#step:1:9.

cf https://github.com/OSGeo/gdal/commit/3231cdfcf44929a6287433695b6a725a020ae6c9 . The issue is related to using the Microsoft unixodbc that comes with MSSQL.

**Pull request merged**